ring/src/lib.rs

```diff
@@ -37,14 +35,22 @@ pub fn find_complement_point<Curve: SWCurveConfig>() -> Affine<Curve> {
     }
 }
 
-// TODO: switch to better hash to curve when available
-pub fn hash_to_curve<A: AffineRepr>(message: &[u8]) -> A {
+// Try and increment hash to curve.
+pub(crate) fn hash_to_curve<F: PrimeField, Curve: SWCurveConfig<BaseField = F>>(message: &[u8]) -> Affine<Curve> {
     use blake2::Digest;
-    use ark_std::rand::SeedableRng;
-
-    let seed = blake2::Blake2s::digest(message);
-    let rng = &mut rand::rngs::StdRng::from_seed(seed.into());
-    A::rand(rng)
+    let mut seed = message.to_vec();
+    let cnt_offset = seed.len();
+    seed.push(0);
+    loop {
+        let hash: [u8; 64] = blake2::Blake2b::digest(&seed[..]).into();
+        let x = F::from_le_bytes_mod_order(&hash);
+        if let Some(point) = Affine::<Curve>::get_point_from_x_unchecked(x, false) {
+            let point = point.clear_cofactor();
```

Comment:
@burdges do we care if it is in the subgroup?

Reply:
I guess he asked the same question here: https://github.com/w3f/ring-proof/pull/31#pullrequestreview-2312086164 

But I guess yes? As the SEED is not in the prime order subgroup, if you add a padding point in the prime subgroup you may end-up with the point at infinity right? Which is not managed in the constraints (and primarily that is why the SEED is chosen to be out of the prime subgroup).

Reply:
Got it.  We cannot compute `x+x` in the SNARK, only `x+y` with `x != y`.  As our snark never checks that `x != y`, we need all our additions to be `x != y` implicitly.

We have two padding points, the start point `hs` and the empty point `he`, maybe equal.  We compute `(..((hs + x[0]) + x[1]) + .. + x[n]) - hs` where the `x[i]` are either a public key if in the bitfield, or zero=infinity if not in the bitfield, or `he` no public key exists there.

We need `-2 hs` to not be a combination of `x[i]`, but `n << group_order`, so either `hs` should be outside the curve, or else we should've proofs of possession for the public keys.